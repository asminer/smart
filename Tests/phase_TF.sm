
// $Id$

print("============================================================\n");
print("\nOz TF  example\nShould look like:\n");
print(" r -> s;   r -> a\n");
print(" s -> r;   s -> a\n\n");
print("-------------------------------------------------------\n");

dtmc oz := {
  state r, n, s;
  init(r:1);
  arcs(
    r:r:2, r:n:1, r:s:1,
    n:r:1, n:s:1,
    s:r:1, s:n:1, s:s:2
  );

  ph int ttnice := phi_TF(potential(in_state(n)));
  ph real error := phr_TF(potential(in_state(n)));
  real pnice := PF(potential(in_state(n)));
};

ph int foo := oz.ttnice;

print_ph(foo);
print("Should be 1: ", oz.pnice, "\n");


print("============================================================\n");
print("\nExample 2\nShould look like:\n");
print(" i -> x;   i -> t\n");
print(" x -> a;\n\n");
print("-------------------------------------------------------\n");

dtmc ex2 := {
  state i, t, x, y, a;
  init(i:1);
  arcs(
    i:t:1, i:x:1,
    x:y:1,
    y:a:1
  );

  ph int tty := phi_TF(potential(in_state(y)));
  ph real error := phr_TF(potential(in_state(y)));
  real py := PF(potential(in_state(y)));
};

print_ph(ex2.tty);
print("Should be 0.5: ", ex2.py, "\n");

print("============================================================\n");
print("\nExample 3\nShould look like:\n");
print(" i -> x;   i -> t\n");
print(" x -> a;\n\n");
print("-------------------------------------------------------\n");

dtmc ex3 := {
  state i, t1, t2, x, y;
  init(i:1);
  arcs(
    i:t1:1, i:x:1,
    t1:t2:1,
    t2:t1:1,
    x:y:1,
    y:t2:1
  );

  ph int tty := phi_TF(potential(in_state(y)));
  ph real tty := phr_TF(potential(in_state(y)));
  real py := PF(potential(in_state(y)));
};

print_ph(ex3.tty);
print("Should be 0.5: ", ex3.py, "\n");


print("============================================================\n");
print("\nExample 4\nShould look like:\n");
print(" x1 -> t;  x1 -> x2;  x1 -> y1\n");
print(" x2 -> t;  x2 -> x3;  x2 -> y2\n");
print(" x3 -> t;  x3 -> a;\n");
print(" y1 -> y2; y2 -> a;\n\n");
print("-------------------------------------------------------\n");

dtmc ex4 := {
  state x1, x2, x3, t1, t2, t3, a1, a2, a3;
  init(x1:1);
  arcs(
    x1:t1:1, x1:a1:1, x1:x2:2,
    x2:t2:1, x2:a2:1, x2:x3:2,
    x3:t3:1, x3:a3:1,
    t1:t2:1, t2:t3:1, t3:t1:1,
    a1:a2:1, a2:a3:1, a3:a2:1
  );

  ph int tt := phi_TF(potential(in_state(a3)));
  ph real error := phr_TF(potential(in_state(a3)));
  real pa3 := PF(potential(in_state(a3)));
};

print_ph(ex4.tt);
print("Should be 0.5: ", ex4.pa3, "\n");

print("============================================================\n");
print("\nExample 5\nShould look like:\n");
print(" x1 -> t;  x1 -> x2;  x1 -> y1\n");
print(" x2 -> t;  x2 -> x3;  x2 -> y2\n");
print(" x3 -> t;  x3 -> a;\n");
print(" y1 -> y2; y2 -> a;\n\n");
print("-------------------------------------------------------\n");

ctmc ex5 := {
  state x1, x2, x3, t1, t2, t3, a1, a2, a3;
  init(x1:1);
  arcs(
    x1:t1:1, x1:a1:1, x1:x2:2,
    x2:t2:1, x2:a2:1, x2:x3:2,
    x3:t3:1, x3:a3:1,
    t1:t2:1, t2:t3:1, t3:t1:1,
    a1:a2:1, a2:a3:1, a3:a2:1
  );

  ph real tt := phr_TF(potential(in_state(a3)));
  ph int error := phi_TF(potential(in_state(a3)));
  real pa3 := PF(potential(in_state(a3)));
};

print_ph(ex5.tt);
print("Should be 0.5: ", ex4.pa3, "\n");

print("============================================================\n");
print("\nThese should generate error messages:\n\n");
print_ph(oz.error);
print_ph(ex2.error);
print_ph(ex3.error);
print_ph(ex4.error);
print_ph(ex5.error);

