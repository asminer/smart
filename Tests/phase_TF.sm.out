============================================================

Oz TF  example
Should look like:
 r -> s;   r -> a
 s -> r;   s -> a

-------------------------------------------------------
Information for phase type tta(oz, ...)
Discrete phase type, state has dimension 1
Reachable states:
State 0: 0
State 1: 2
State 2: a
3 states total
Accepting state index: 2
No trap state
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 0 with weight 0.5
	To state 1 with weight 0.25
	To state 2 with weight 0.25
From state 1:
	To state 0 with weight 0.25
	To state 1 with weight 0.5
	To state 2 with weight 0.25
From state 2:
	To state 2 with weight 1
7 edges total
End of information for phase type

Should be 1: 1
============================================================

Example 2
Should look like:
 i -> x;   i -> t
 x -> a;

-------------------------------------------------------
Information for phase type tta(ex2, ...)
Discrete phase type, state has dimension 1
Reachable states:
State 0: 0
State 1: 1
State 2: t
State 3: a
4 states total
Accepting state index: 3
Trap state index: 2
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 0.5
	To state 2 with weight 0.5
From state 1:
	To state 3 with weight 1
From state 2:
	To state 2 with weight 1
From state 3:
	To state 3 with weight 1
5 edges total
End of information for phase type

Should be 0.5: 0.5
============================================================

Example 3
Should look like:
 i -> x;   i -> t
 x -> a;

-------------------------------------------------------
Information for phase type tta(ex3, ...)
Discrete phase type, state has dimension 1
Reachable states:
State 0: 0
State 1: 1
State 2: t
State 3: a
4 states total
Accepting state index: 3
Trap state index: 2
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 0.5
	To state 2 with weight 0.5
From state 1:
	To state 3 with weight 1
From state 2:
	To state 2 with weight 1
From state 3:
	To state 3 with weight 1
5 edges total
End of information for phase type

Should be 0.5: 0.5
============================================================

Example 4
Should look like:
 x1 -> t;  x1 -> x2;  x1 -> y1
 x2 -> t;  x2 -> x3;  x2 -> y2
 x3 -> t;  x3 -> a;
 y1 -> y2; y2 -> a;

-------------------------------------------------------
Information for phase type tta(ex4, ...)
Discrete phase type, state has dimension 1
Reachable states:
State 0: 0
State 1: 1
State 2: 3
State 3: 2
State 4: 7
State 5: t
State 6: a
7 states total
Accepting state index: 6
Trap state index: 5
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 0.5
	To state 2 with weight 0.25
	To state 5 with weight 0.25
From state 1:
	To state 3 with weight 0.5
	To state 4 with weight 0.25
	To state 5 with weight 0.25
From state 2:
	To state 4 with weight 1
From state 3:
	To state 5 with weight 0.5
	To state 6 with weight 0.5
From state 4:
	To state 6 with weight 1
From state 5:
	To state 5 with weight 1
From state 6:
	To state 6 with weight 1
12 edges total
End of information for phase type

Should be 0.5: 0.5
============================================================

Example 5
Should look like:
 x1 -> t;  x1 -> x2;  x1 -> y1
 x2 -> t;  x2 -> x3;  x2 -> y2
 x3 -> t;  x3 -> a;
 y1 -> y2; y2 -> a;

-------------------------------------------------------
Information for phase type tta(ex5, ...)
Continuous phase type, state has dimension 1
Reachable states:
State 0: 0
State 1: 1
State 2: 3
State 3: 2
State 4: 7
State 5: t
State 6: a
7 states total
Accepting state index: 6
Trap state index: 5
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 2
	To state 2 with weight 1
	To state 5 with weight 1
From state 1:
	To state 3 with weight 2
	To state 4 with weight 1
	To state 5 with weight 1
From state 2:
	To state 4 with weight 1
From state 3:
	To state 5 with weight 1
	To state 6 with weight 1
From state 4:
	To state 6 with weight 1
From state 5:
From state 6:
10 edges total
End of information for phase type

Should be 0.5: 0.5
============================================================

These should generate error messages:

ERROR in file phase_TF.sm near line 20:
    Underlying distribution is ph int instead of ph real
null phase type
ERROR in file phase_TF.sm near line 46:
    Underlying distribution is ph int instead of ph real
null phase type
ERROR in file phase_TF.sm near line 71:
    Underlying distribution is ph int instead of ph real
null phase type
ERROR in file phase_TF.sm near line 99:
    Underlying distribution is ph int instead of ph real
null phase type
ERROR in file phase_TF.sm near line 126:
    Underlying distribution is ph real instead of ph int
null phase type
